tests/fixtures/descriptor_tests/with_create_descriptor_hook/core/hooks/descriptor_operations.py

```diff
@@ -0,0 +1,128 @@
+        pc = self.shotgun.find_one(
+            "PipelineConfiguration",
+            [["id", "is", self.pipeline_configuration_id]],
+            ["sg_external_site_credentials"]
+        )
+
+        # If we've found something, create a connection to that remote site,
+        # but do not connect right away.
+        if pc and pc.get("sg_external_site_credentials"):
+            self._remote_sg = Shotgun(connect=False, **pc["sg_external_site_credentials"])
+        else:
+            self._remote_sg = None
+
+    def download_local(self, descriptor):
+        """
+        This method is invoked during bootstrapping when downloading the core or
+        the bundles in the environment from the configuration.
+
+        Your method must download the descriptor's content before the method
+        ends of the bootstrap manager will report an error.
+
+        The safest way to download a bundle into the bundle cache is by using the
+        :meth:`sgtk.descriptor.Descriptor.external_download` context manager.
+
+        Any exception raised by this code will abort the bootstrap process.
+
+        :param descriptor: The descriptor that will be downloaded.
+        :type descriptor: :class:`~sgtk.descriptor.Descriptor`
+        """
+        # First try to download from the local cache.
+        if self._download_from_sg_cache(self.shotgun, descriptor):
+            return
+
+        # Then try to download the from remote cache.
+        if self._remote_sg and self._download_from_sg_cache(self._remote_sg, descriptor):
+            return
+
+        # Give up, we're going to talk to the source.
+        descriptor.download_local()
+
+    def _download_from_sg_cache(self, shotgun, descriptor):
+        """
+        Downloads a bundle from a Shotgun site.
+
+        :param shotgun: Connection to the site to download from.
+        :type shotgun: :class:`shotgun_api3.Shotgun`
+
+        :param descriptor: Descriptor that needs to be downloaded.
+        :type descriptor: :class:`~sgtk.descriptor.Descriptor`
+
+        :returns: Returns ``True`` is something was downloaded.
+        :rtype: bool
+        """
+
+        # Get the uri form the descriptor. We'll have to make sure that descriptor values
+        # after the ? are always sorted the same or this will be an issue.
+        entity = shotgun.find_one(
+            "CustomNonProjectEntity01",
+            [["code", "is", descriptor.get_uri()]], ["sg_uploaded_bundle"]
+        )
+        if not entity:
+            return False
+
+        # When calling external_download, the method yields to us
+        # a path that needs to be filled with the files. If the
+        # with ends normally, the files are copied into the cache.
+        # In an exception is raised, the files are deleted and the exception
+        # bubbles upward.
+        with descriptor.external_download() as external_path:
+            download_and_unpack_attachment(shotgun, entity["sg_uploaded_bundle"], external_path)
```

Comment:
i guess this is a safe assumption to make - that the descriptor api is ok to open this up for external processing?

Reply:
Maybe the best approach would be to provide a base class to the hook with a method that invokes the internal API?